api/src/wfl/module/staged.clj

```diff
@@ -125,17 +125,18 @@
     (if-not started (start workload (utc-now)) workload)))
 
 (defn ^:private update-staged-workload
-  "Use transaction `tx` to update `workload` statuses."
-  [tx {:keys [started finished] :as workload}]
+  "Update `workload` stages."
+  [{:keys [started finished] :as workload}]
   (letfn [(update! [{:keys [id source executor sink] :as workload} now]
             (-> workload
                 (source/update-source!)
                 (executor/update-executor!)
                 (sink/update-sink!))
-            (patch-workload tx workload {:updated now})
-            (when (every? stage/done? [source executor sink])
-              (patch-workload tx workload {:finished now}))
-            (workloads/load-workload-for-id tx id))]
+            (jdbc/with-db-transaction [tx (postgres/wfl-db-config)]
+              (patch-workload tx workload {:updated now})
+              (when (every? stage/done? [source executor sink])
+                (patch-workload tx workload {:finished now}))
+              (workloads/load-workload-for-id tx id)))]
     (if (and started (not finished)) (update! workload (utc-now)) workload)))
```

Comment:
It isn't new in this PR but noticed that update-sink! and patch-workload run in separate TXs?
I guess workload state will converge. (Batch updates are simpler here.)

Reply:
Yes, all of the individual stages defined transactions as needed to avoid external calls within open transactions.

But yes, there are thread safety concerns where retries are concerned:

https://broadinstitute.atlassian.net/browse/GH-1423